File: src/C/llt.c
#include "chompack.h"

void llt(const int_t n,         // order of matrix
	 const int_t nsn,       // number of supernodes/cliques
	 const int_t *snpost,   // post-ordering of supernodes
	 const int_t *snptr,    // supernode pointer
	 const int_t *relptr, 
	 const int_t *relidx, 
	 const int_t *chptr, 
	 const int_t *chidx,
	 const int_t *blkptr, 
	 double * restrict blkval,
	 double * restrict fws,  // frontal matrix workspace
	 double * restrict upd,  // update matrix workspace
	 int_t * restrict upd_size  
	 ) {

  int nn,na,nj,offset,i,j,k,ki,l,N,nup=0;
  double * restrict U;
  double dOne=1.0,dZero=0.0;
  char cL='L',cR='R',cN='N';

  U = upd;   // pointer to top of update storage

  for (ki=0;ki<nsn;ki++) {
    k = snpost[ki];
    nn = snptr[k+1]-snptr[k];
    na = relptr[k+1]-relptr[k];
    nj = na + nn;        

    // compute [I; L_{Ak,Nk}]*D_k*[I;L_{Ak,Nk}]'; store in frontal workspace
    dtrmm_(&cR, &cL, &cN, &cN, &na, &nn, &dOne, blkval+blkptr[k], &nj, blkval+blkptr[k]+nn, &nj);
    dsyrk_(&cL, &cN, &nj, &nn, &dOne, blkval+blkptr[k], &nj, &dZero, fws, &nj);
    
    // add update matrices to frontal matrix
    for (l=chptr[k+1]-1;l>=chptr[k];l--) {
      nup--;
      U -= upd_size[nup]*upd_size[nup];
      // extend-add
      offset = relptr[chidx[l]];
      N = relptr[chidx[l]+1] - offset;
      for (j=0; j<N; j++) {
	for (i=j; i<N; i++)
	  fws[nj*relidx[offset+j]+relidx[offset+i]] += U[N*j+i];
      }
    }

    // if supernode k is not a root node, push update matrix onto stack
    if (na > 0) {   
      upd_size[nup++] = na;
      dlacpy_(&cL, &na, &na, fws+nn*nj+nn, &nj, U, &na);
      U += na*na;
    }

    // copy the leading nn columns of frontal matrix to blkval
    dlacpy_(&cL, &nj, &nn, fws, &nj, blkval+blkptr[k], &nj);
  }

  return;
}

File: src/python/pybase/llt.py
from cvxopt import matrix, blas, lapack
from chompack.symbolic import cspmatrix
from chompack.misc import frontal_add_update

def llt(L):
    """
    Supernodal multifrontal Cholesky product:

    .. math::
         X = LL^T

    where :math:`L` is lower-triangular. On exit, the argument `L`
    contains the product :math:`X`.

    :param L:    :py:class:`cspmatrix` (factor)
    """

    assert isinstance(L, cspmatrix) and L.is_factor is True, "L must be a cspmatrix factor"

    n = L.symb.n
    snpost = L.symb.snpost
    snptr = L.symb.snptr
    chptr = L.symb.chptr
    chidx = L.symb.chidx

    relptr = L.symb.relptr
    relidx = L.symb.relidx
    blkptr = L.symb.blkptr
    blkval = L.blkval

    stack = []

    for k in snpost:

        nn = snptr[k+1]-snptr[k]       # |Nk|
        na = relptr[k+1]-relptr[k]     # |Ak|
        nj = na + nn

        # compute [I; L_{Ak,Nk}]*D_k*[I;L_{Ak,Nk}]'; store in F
        blas.trmm(blkval, blkval, side = "R", m = na, n = nn, ldA = nj,\
                  ldB = nj, offsetA = blkptr[k], offsetB = blkptr[k]+nn)
        F = matrix(0.0, (nj, nj))
        blas.syrk(blkval, F, n = nj, k = nn, offsetA = blkptr[k], ldA = nj)

        # if supernode k has any children, subtract update matrices 
        for _ in range(chptr[k],chptr[k+1]):
            Ui, i = stack.pop()
            frontal_add_update(F, Ui, relidx, relptr, i)

        # if supernode k is not a root node, push update matrix onto stack
        if na > 0:
            Uk = matrix(0.0,(na,na))
            lapack.lacpy(F, Uk, m = na, n = na, uplo = 'L', offsetA = nn*nj+nn, ldA = nj)
            stack.append((Uk,k))

        # copy leading Nk columns of F to blkval
        lapack.lacpy(F, blkval, m = nj, n = nn, ldA = nj, uplo = 'L',\
                     ldB = nj, offsetB = blkptr[k])

    L._is_factor = False

    return 
